policy: Fix Distillery Tests

nathanjsweet opened the pull request:
See commits for details.

squeed commented on pkg/policy/distillery_test.go:
```diff
@@ -1624,12 +1625,12 @@ var (
 					{
 						Port:     "64",
 						EndPort:  127,
-						Protocol: api.ProtoAny,
```
Why are we changing the test data here?

nathanjsweet replied:
To reduce the overall test data before. The test was actually failing before and required a lot more entries than was in the "expected" field. Changing this to TCP reduces the overall test data without sacrificing a comprehensive test of the logic.

squeed commented on pkg/policy/distillery_test.go:
```diff
@@ -1663,72 +1669,83 @@ func Test_EnsureDeniesPrecedeAllows(t *testing.T) {
 		{"deny_world_no_labels", api.Rules{ruleL3DenyWorld, ruleL3AllowWorldIP}, newMapState(map[Key]MapStateEntry{
-			mapKeyL3WorldIngress: mapEntryDeny,
-			mapKeyL3WorldEgress:  mapEntryDeny,
+			mapKeyL3WorldIngress:  mapEntryDeny,
```
Why are there now 2 keys for the same direction? Shouldn't a single wildcard be sufficient?

nathanjsweet replied:
Other CIDR identities are automatically selected by the reserved:World identity by the policy repository. This test was silently failing before, nothing in the port range work caused this change. We can and should address/understand these redundancies, but not in this PR.

squeed commented on pkg/policy/distillery_test.go:
```diff
@@ -1663,72 +1669,83 @@ func Test_EnsureDeniesPrecedeAllows(t *testing.T) {
 			mapKeyL3SubnetIngress: mapEntryDeny,
 			mapKeyL3SubnetEgress:  mapEntryDeny,
+		})}, {"deny_one_ip_with_a_larger_subnet", api.Rules{ruleL3DenySubnet, ruleL3AllowWorldIP}, newMapState(map[Key]MapStateEntry{
+			mapKeyL3SubnetIngress:        mapEntryDeny,
```
same question here -- why are there new mapstate entries here?

nathanjsweet replied:
I'm not sure what change caused this, but I'm confident it was not the port range changes.

jrajahalme commented on pkg/maps/policymap/policymap.go:
```diff
@@ -117,6 +117,11 @@ func (k *PolicyKey) GetDestPort() uint16 {
 
+// GetPortMask returns the port mask of the key
+func (k *PolicyKey) GetPortMask() uint16 {
+	return 0xffff
```
Should this be conditional on `DestPortNetwork` being 0?

nathanjsweet replied:
Eventually. The pull request that finally introduces port ranges will add logic here.

nathanjsweet committed:
policy: Fix InvertedPortMask Logic

Inverted port masks were introduced in a previous
commit, but they did not take into account all of the
wildcard policy keys that are hardcoded both in tests
and, unfortunately, in production code. This fixes
all InvertedPortMask logic so that wild cards are
correctly instantiated and accounted for.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Fix Proxy Port Tests

The default proxy port is 1. The dummy proxy
port value of 4242 should no longer be the default.
Addtionally, almost all distillery tests have had
their comparison logic fixed to compare mapstates.
The last distillery test to be fixed will be fixed
in another commit.

Co-authored-by: Jarno Rajahalme <jarno@isovalent.com>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Fix Deny Precedence Distillery Tests

Deny precedence tests have not been using correct
comparison logic for mapstates. This updates it
to use the correct logic. It makes the named-port
sub-test and port-range tests more legible.
Some test data has been renamed because it was
incorrectly named.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Fix Rare Deny Merge Bug

While fixing the distillery tests I noticed that the
test "broad_allow_is_a_portproto_subset_of_a_specific_deny"
was failing intermittently. This revealed that in very rare
circumstances the policy engine would merge a deny entry
into a pre-existing allow that was set to be deleted completely
(not just have an ownership entry removed).

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

gentoo-root commented on pkg/policy/types/types.go:
```diff
@@ -121,26 +121,10 @@ func (k Key) PortIsEqual(c Key) bool {
 
-// maskPrefixMap is the map of all the possible portMasks and their
-// associated prefix lengths.
-var maskPrefixMap = map[uint16]uint{
```
The old code returned the number of leading zeroes, the new code switched to trailing. Is it intended? Was the old code wrong?

nathanjsweet replied:
The logic is correct, the name is poor.

gentoo-root commented on pkg/policy/mapstate.go:
```diff
@@ -861,21 +867,25 @@ func (ms *mapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapState
 			return true
 		})
+		for _, delete := range deletes {
+			if !delete.Add {
```
Seems like the `!delete.Add` and `update.Add` checks are not strictly needed.

nathanjsweet replied:
The are not.
